### DISCUSSION TITLE
Embedded GUI Using Linux Frame Buffer Device with Slint UI and std

### DISCUSSION THREAD

**Discussion by dongjian** [slint-ui/slint#2832] [Q&A] [answered]
Is there a way to render the slint ui directly on linux's /dev/fb0 while using rust's std ? I looked at the documentation, but couldn't find a tutorial or demo . please help

**Comment by ogoffart** [comment, marked as answer]
Not out of the box ( https://github.com/slint-ui/slint/issues/712 )

But this is possible if you use a custom platform and the software renderer. (unfortunately not the GL renderer at this point)

The documentation on how to create a custom platform is here: https://docs.rs/slint/latest/slint/docs/mcu/index.html 
This talks about no_std environment, but you can also use it with std environment (with the std feature of slint)
You'll have to use some library to draw the image on the frame buffer. Maybe there is a crate for that.

This would make for a good demo actually.

**Reply by nilclass** [reply]
I was curious about this too, so I gave it a try (using the `linuxfb` crate I wrote a couple of years ago).

Here is a small example: https://github.com/nilclass/slint-framebuffer-example
It's not a full-blown demo (e.g. it doesn't even support any input), but could serve as a starting point for one.

**Reply by tronical** [reply]
Wow, this is very cool. Very nice to see how few lines of code it is to get those pixels onto the screen.

**Comment by igiona** [comment]
Is there any update on this?

The example provided above works for me as well, but it sounds like quite experimental rather than a production ready solution.
Wouldn't `linuxkms` be an option? If yes, how can be setup?

Please apologize if this should be completely wrong, I'm still trying to pull all pieces of the puzzle together with linux-fb, skia, femto, winit, wayland, qt, openGL, kms, drm and so on and so forth...

Thanks!

**Reply by tronical** [reply]
Hi! Support for basic linuxfb rendering in combination with the Slint software renderer was added in Slint 1.7.0. It should be sufficient to enable backend-linuxkms and renderer-software. It should get selected automatically if no other windowing system was found or GPU.

**Reply by igiona** [reply]
@tronical thank you for the update, this is good news!

I tried to cross-compile my simple slint app with the following Cargo config:

```toml
slint = { version = "1.8.0", default-features = false, features = ["std", "accessibility", "compat-1-2", "backend-linuxkms-noseat", "renderer-software"] }
```

but I get a linker error regarding `libinput`:

```
/../lib/gcc/arm-none-linux-gnueabihf/10.3.1/../../../../arm-none-linux-gnueabihf/bin/ld: cannot find -linput
          collect2: error: ld returned 1 exit status
```

From what I can see, this library is related to wayland (https://wayland.freedesktop.org/libinput/doc/latest/)?
Is there an option to opt it out? ..and anyhow I don't need touch/keyboard/mouse support for now ;)

**Reply by tronical** [reply]
At the moment this library is mandatory. It's used to read touch/mouse/key events. Fortunately it doesn't depend on Wayland. If you're using cross, you could use this cross config to compile within a container that has libinput:

https://github.com/slint-ui/slint/blob/master/Cross.toml

**Reply by igiona** [reply]
I see, it would be nice to be able to opt-it out for systems that do not have inputs.
I'm not using cross, I build against an SDK built via Yocto/BitBake.
Using cross is not an option for me, since our `libc` will likely not match.

I will try to add the `libinput` recipe, but I'm afraid it will pull in a lot of dependencies: [recipes-graphics](http://cgit.openembedded.org/openembedded-core/tree/meta/recipes-graphics/wayland/libinput_1.16.4.bb?h=hardknott)
I'm not so sure if it is a sound thing to do.

**Reply by tronical** [reply]
Yes, agreed. It would be nice to offer a build option without libinput.

**Reply by igiona** [reply]
@tronical adding that recipe didn't do too much damages on the image size😄 

Now though, I'm getting other linker errors now:

```bash
  = note: /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find Scrt1.o: No such file or directory
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find crti.o: No such file or directory
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find crtbeginS.o: No such file or directory
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find -lgcc_s
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find /lib/libc.so.6
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find /usr/lib/libc_nonshared.a
          /home/igiona/sdks/lcd_sdk2/sysroots/x86_64-scewo-linux/usr/libexec/arm-scewo-linux-gnueabi/gcc/arm-scewo-linux-gnueabi/10.3.0/real-ld: cannot find /lib/ld-linux-armhf.so.3
```

All these files and libs are actually available, somehow are not found ?!
What's pretty weird to me, is that the linker wants to stuff in `/lib` rather than looking in the provide `-L` paths...
I've "fixed" them with:
```
sudo ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/lib/libc.so.6 /lib/libc.so.6
sudo ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/libc_nonshared.a /usr/lib/libc_nonshared.a
sudo ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/lib/ld-linux-armhf.so.3 /lib/ld-linux-armhf.so.3
```

and so these 3 are gone.

EDIT: 
~ With regards with `-lgcc_s` I'm not sure how to address it... ~
I was able to workaroun this as follow:
```
 ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/lib/libgcc_s.so /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/
 ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/lib/libgcc_s.so.1 /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/
 ln -s /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/arm-scewo-linux-gnueabi/10.3.0/libgcc.a /home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/
ln -s /home/impy/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/lib/libc.so.6 /home/impy/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-scewo-linux-gnueabi/usr/lib/
```

I'm left with the object files errors.
`Scrt1.o`, `crti.o` are actually already in one of the -L path, and `crtbeginS.o` is in the same folder as the linker, so I've no clue why is not found?!

Any clue?

**Reply by tronical** [reply]
These object files and libraries are usually found in your sysroot. How did you pass the sysroot to the linker?

**Reply by igiona** [reply]
> These object files and libraries are usually found in your sysroot. How did you pass the sysroot to the linker?

In the meanwhile, I've been able to build 🎉 exactly by passing the sysroot via config.toml to the linker:

```toml
[env]
PKG_CONFIG_SYSROOT_DIR="/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi"
PKG_CONFIG_PATH="/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi/usr/lib/pkgconfig:/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi/usr/share/pkgconfig"

[target.armv7-unknown-linux-gnueabihf]
linker="/home/igiona/sdks/lcd_sdk2/sysroots/x86_64-org-linux/usr/bin/arm-org-linux-gnueabi/arm-org-linux-gnueabi-gcc"
rustflags = [
    "-C", "link-arg=--sysroot=/home/igiona/sdks/lcd_sdk2/sysroots/armv7vet2hf-neon-org-linux-gnueabi",
]
```

**Reply by tronical** [reply]
Excellent, that looks like the correct solution for your setup :)
